### ISSUE TITLE
linspace enpoint overflow for large stop values with int64

### ISSUE THREAD

**Issue by PhilippThoelke** [numpy/numpy#17155]
When calling `np.linspace()` with very large stop values (e.g. `np.iinfo(np.int64).max`) and `dtype=np.int64`, an overflow occurs in the cast from the intermediate float64 array to the result int64 array.

### Reproducing code example:

```python
import numpy as np
print(np.linspace(0, np.iinfo(np.int64).max, 2, dtype=np.int64))
```
This produces `[                   0 -9223372036854775808]` while the expected endpoint is expected to be `9223372036854775807` (i.e. `np.iinfo(np.int64).max`).

### Error message:
None.

### Numpy/Python version information:

1.19.1 3.7.8 (default, Jul  9 2020, 07:05:14)
[GCC 10.1.0]

**Comment by PhilippThoelke** [comment]
A potential fix for this would be to first cast the intermediate float64 array to the output dtype and then setting the last array element to the original stop value, similar to what is already being done:
https://github.com/numpy/numpy/blob/529c61bdf69356b4a428d48bb5bc72c2417ed0cf/numpy/core/function_base.py#L163-L164

**Comment by eric-wieser** [comment]
That wouldn't really solve the problem - precision is still lost for
```
In [101]: np.linspace(0, np.iinfo(np.int64).max, 4, dtype=np.int64)[2]
Out[101]: 6148914691236516864

# expected
In [102]: np.iinfo(np.int64).max * 2// 3
Out[102]: 6148914691236517204
```

**Comment by PhilippThoelke** [comment]
That is true but at least this fixes the overflow in the cast of the last array element, causing it to be `np.iinfo(np.int64).min` instead of `np.iinfo(np.int64).max`. In my opinion this already helps, even though other elements might still be imprecise. When using the function, I would expect the first element to definitely be `start` and the last one to be `stop`.

Fixing the precision problem here requires some more work, you are right with that.

**Comment by eric-wieser** [comment]
Even fixing out-of-bounds values is hard here:
```
In [122]: np.linspace(np.iinfo(np.int64).max-2**9+1, np.iinfo(np.int64).max, 2, dtype=np.int64)
Out[122]: array([-9223372036854775808, -9223372036854775808], dtype=int64)
```

**Comment by PhilippThoelke** [comment]
The last element of the result in your example would be correct with my proposal (setting it to `stop` after casting) and one could easily also set the first element to `start`. But you are right, this does not fix the problem and values between the first and last element might still exhibit overflows as in this case for example:
```python
>>> import numpy as np
>>> np.linspace(np.iinfo(np.int64).max-2**9+1, np.iinfo(np.int64).max, 3, dtype=np.int64)
array([ 9223372036854775296, -9223372036854775808,  9223372036854775807])
```
Note: this is the modified version of `np.linspace()` where the first and last element of the result are set to the original `start` and `stop` values respectively after casting from the intermediate `float64` representation to `dtype` (`int64` in this example).

One way to completely get around this would of course be to use `float128` for the intermediate computations whenever `dtype` is `int64`. Would that be an option or is such a workaround an undesired solution?

**Comment by eric-wieser** [comment]
> One way to completely get around this would of course be to use `float128` for the intermediate computations whenever `dtype` is `int64`. Would that be an option

Not on Windows, where `longdouble` has no extra precision.

**Comment by m-newhauser** [comment]
Is this still an issue?

**Comment by seberg** [comment]
If anyone wants to pick this up, make sure to either get maintainer buy-in or explain how this affects real world code.
